core/trino-main/src/main/java/io/trino/json/ir/TypedValue.java

```diff
@@ -0,0 +1,140 @@
+        this.doubleValue = doubleValue;
+        this.booleanValue = false;
+    }
+
+    public TypedValue(Type type, boolean booleanValue)
+    {
+        requireNonNull(type, "type is null");
+        checkArgument(boolean.class.equals(type.getJavaType()), "boolean value does not match the type %s", type);
+
+        this.type = type;
+        this.objectValue = null;
+        this.longValue = 0L;
+        this.doubleValue = 0e0;
+        this.booleanValue = booleanValue;
+    }
+
+    public static TypedValue fromValueAsObject(Type type, Object valueAsObject)
+    {
+        if (long.class.equals(type.getJavaType())) {
+            checkState(valueAsObject instanceof Long, "%s value does not match the type %s", valueAsObject.getClass(), type);
+            return new TypedValue(type, (long) valueAsObject);
+        }
+        if (double.class.equals(type.getJavaType())) {
+            checkState(valueAsObject instanceof Double, "%s value does not match the type %s", valueAsObject.getClass(), type);
+            return new TypedValue(type, (double) valueAsObject);
+        }
+        if (boolean.class.equals(type.getJavaType())) {
+            checkState(valueAsObject instanceof Boolean, "%s value does not match the type %s", valueAsObject.getClass(), type);
+            return new TypedValue(type, (boolean) valueAsObject);
+        }
+        checkState(type.getJavaType().isAssignableFrom(valueAsObject.getClass()), "%s value does not match the type %s", valueAsObject.getClass(), type);
+        return new TypedValue(type, valueAsObject);
+    }
+
+    public Type getType()
+    {
+        return type;
+    }
+
+    public Object getObjectValue()
+    {
+        checkArgument(objectValue != null, "the type %s is represented as %s. call another method to retrieve the value", type, type.getJavaType());
+        checkArgument(type.getJavaType().isAssignableFrom(objectValue.getClass()), "%s value does not match the type %s", objectValue.getClass(), type);
+        return objectValue;
+    }
+
+    public long getLongValue()
+    {
+        checkArgument(long.class.equals(type.getJavaType()), "long value does not match the type %s", type);
+        return longValue;
+    }
+
+    public double getDoubleValue()
+    {
+        checkArgument(double.class.equals(type.getJavaType()), "double value does not match the type %s", type);
+        return doubleValue;
+    }
+
+    public boolean getBooleanValue()
+    {
+        checkArgument(boolean.class.equals(type.getJavaType()), "boolean value does not match the type %s", type);
+        return booleanValue;
+    }
+
+    public Object getValueAsObject()
+    {
+        if (long.class.equals(type.getJavaType())) {
+            return longValue;
+        }
+        if (double.class.equals(type.getJavaType())) {
+            return doubleValue;
+        }
+        if (boolean.class.equals(type.getJavaType())) {
+            return booleanValue;
+        }
+        return objectValue;
```

Comment:
If that is called often you can set the "object" value to appropriate value already in the consturctor.

Reply:
I set "valueAsObject" in the constructor.

Do you think it would be good to get rid of the 4 optional fields (Object, long, double, boolean), and keep only the one value? In that case I'd have to add a cast on every call to `getObjectValue`, `getLongValue`, 'getDoubleValue` and `getBooleanValue`.